NDTensors/src/dense/dense.jl

```diff
@@ -3,70 +3,101 @@
 #
 using LinearAlgebra: BlasFloat
 
-struct Dense{ElT,VecT<:AbstractArray} <: TensorStorage{ElT}
-  data::VecT
-  function Dense{ElT,VecT}(data::AbstractArray) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(data)
+struct Dense{ElT,DataT<:AbstractArray} <: TensorStorage{ElT}
+  data::DataT
+  function Dense{ElT,DataT}(data::AbstractArray) where {ElT,DataT<:AbstractArray{ElT}}
+    return new{ElT,DataT}(data)
   end
 
-  # Special case of Vector
-  function Dense{ElT,VecT}(data::Vector) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(data)
-  end
+  # # Special case of Vector
+  # function Dense{ElT,DataT}(data::Vector) where {ElT,DataT<:AbstractArray{ElT}}
+  #   return new{ElT,DataT}(data)
+  # end
 
   # Special case of Array which gets turned into a Vector
-  function Dense{ElT,VecT}(data::Array) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(vec(data))
-  end
+  # function Dense{ElT,DataT}(data::Array) where {ElT,DataT<:AbstractArray{ElT}}
+  #   return new{ElT,DataT}(vec(data))
+  # end
+end
+
+#Start with high information constructors and move to low information constructors
+function Dense{ElT,DataT}() where {ElT,DataT<:AbstractArray{ElT}}
+  return Dense{ElT,DataT}(DataT())
+end
+
+# Construct from a set of indices
+# This will fail if zero(ElT) is not defined for the ElT
+function Dense{ElT,DataT}(inds::Tuple) where {ElT,DataT<:AbstractArray{ElT}}
+  return Dense{ElT,DataT}(zeros(DataT, dim(inds)))
+end
+
+function Dense{ElT,DataT}(
+  ::UndefInitializer, inds::Tuple
+) where {ElT,DataT<:AbstractArray{ElT}}
+  return Dense{ElT,DataT}(DataT(undef, dim(inds)))
+end
+
+function Dense{DataT}() where {DataT<:AbstractArray}
+  return default_storagetype(DataT)()
+end
+
+function Dense{DataT}(inds::Tuple) where {DataT<:AbstractArray}
+  return default_storagetype(DataT)(inds)
 end
 
-function Dense(data::VecT) where {VecT<:AbstractArray{ElT}} where {ElT}
-  return Dense{ElT,VecT}(data)
+function Dense{DataT}(::UndefInitializer, inds::Tuple) where {DataT<:AbstractArray}
+  return default_storagetype(DataT)(undef, inds)
 end
 
-function Dense(data::Array{ElT}) where {ElT}
-  return Dense{ElT,Vector{ElT}}(vec(data))
+function Dense{DataT}(x::Number, dim::Integer) where {DataT<:AbstractArray}
+  return Dense(fill!(DataT{typeof(x)}(undef, dim), x))
```

Comment:
In general you should avoid using `DataT{...}`, i.e. explicitly listing type parameters of an unknown type with that notation, since that won't be generic to all data types, for example if they have other type parameters like a tensor order or buffer type. Instead use generic functions, i.e.: `similartype(DataT, typeof(x))` (or `adapt_type` or `set_eltype` or whatever we ultimately decide is the best practice).

Reply:
Also as I lay out in my comment below, I don't think we should ever treat the first type parameter as the data type (it should always be interpreted as the element type).